### PR TITLE
feat(EMS-3265-3201-3270-3271): iterations - Account - Name fields validation

### PR DESCRIPTION
**Pull request by Zainzzkk**
## Introduction :pencil2:

This PR introduces name validation for create account name fields for EMS-3265-3201-3270-3271

## Resolution :heavy_check_mark:

* Changed to nameValidation for create account details, loss payee name, name on policy position and agent details name
* Added new regexValidation function
* Removed replication by required validation functions using regexValidation
* Added `alphaCharactersHyphenAndApostropheOnlyValidation` for name fields
* Added regex pattern for alphabet, apostrophes, spaces and hyphens
* Added test
* Added shared name e2e test assertion

### CODE REVIEW

File: e2e-tests/constants/examples/name-examples.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+export const INVALID_NAMES = {
+  SPECIAL_CHARACTERS: 'Name!?',
+  SPECIAL_CHARACTERS_SPACE: 'Mock name!',
```
Comment (ttbarnes) [inline e2e-tests/constants/examples/name-examples.js:3]: `_AND_SPACE`?

File: e2e-tests/shared-test-assertions/name-field-validation/index.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+ * @param {Integer} errorIndex: Index of the summary list error
+ * @param {Object} errorMessages: Email error messages
+ * @param {Integer} totalExpectedErrors: Total expected errors in the form
+ * @param {Integer} totalExpectedOtherErrorsWithName: Total expected errors in the form when name is valid.
+ * @param {Boolean} shouldHaveOtherErrors: If other error messages should be present on the page if fieldId is valid.
+ * @returns {Function} Mocha describe block with assertions.
+ */
+export const assertNameFieldValidation = ({
+  fieldId,
+  nameMaximumCharacters,
```
Comment (ttbarnes) [inline e2e-tests/shared-test-assertions/name-field-validation/index.js:18]: could just be `maximum`? Like in `percentageFieldValidation` we have `minimum`

File: e2e-tests/insurance/cypress/e2e/journeys/account/create/your-details/account-create-your-details-validation.spec.js [hunk 3]
```diff
@@ -49,24 +50,24 @@ context('Insurance - Account - Create - Your details page - empty form validatio
...
-      expectedErrorsCount: TOTAL_REQUIRED_FIELDS,
-      expectedErrorMessage: YOUR_DETAILS_ERROR_MESSAGES[LAST_NAME].IS_EMPTY,
-    });
+  assertNameFieldValidation({
+    fieldId: LAST_NAME,
+    errorIndex: 1,
+    nameMaximumCharacters,
+    errorMessages: YOUR_DETAILS_ERROR_MESSAGES[LAST_NAME],
+    totalExpectedErrors: 4,
+    totalExpectedOtherErrorsWithName: 3,
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/account/create/your-details/account-create-your-details-validation.spec.js:70]: Could have a `const` for `totalExpectedErrors` and `totalExpectedErrors`?

File: e2e-tests/insurance/cypress/e2e/journeys/export-contract/agent-details/validation/agent-details-validation.spec.js [hunk 4]
```diff
@@ -52,40 +51,19 @@ context('Insurance - Export contract - Agent details page - form validation', ()
...
-      const value = 'a1!';
-
-      cy.submitAndAssertFieldErrors({ ...assertions, value, expectedErrorMessage: ERROR_MESSAGES_OBJECT.INCORRECT_FORMAT });
+    assertNameFieldValidation({
+      fieldId: NAME,
+      errorIndex: 0,
+      nameMaximumCharacters: 'a'.repeat(MAXIMUM_CHARACTERS.AGENT_NAME + 1),
+      errorMessages: ERROR_MESSAGES_OBJECT,
+      totalExpectedErrors: expectedErrorsCount,
+      totalExpectedOtherErrorsWithName: 2,
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/export-contract/agent-details/validation/agent-details-validation.spec.js:66]: `WithValidName`?

File: e2e-tests/shared-test-assertions/name-field-validation/index.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+    it(`should render a validation error the ${fieldId} contains numbers`, () => {
+      cy.submitAndAssertFieldErrors({
+        ...assertions,
+        value: INVALID_NAMES.NUMBERS,
+        expectedErrorMessage: incorrectFormatErrorMessage,
+      });
+    });
+
+    it(`should render a validation error the ${fieldId} is above the maximum number of characters`, () => {
```
Comment (ttbarnes) [inline e2e-tests/shared-test-assertions/name-field-validation/index.js:72]: This and the above tests are missing a "when" between "error the" 😛

File: src/ui/server/constants/regex.ts [hunk 6]
```diff
@@ -5,7 +5,17 @@ export const REGEX = {
...
+
+  /**
+   * ALPHA_CHARACTERS_AND_SPACE_HYPHEN_APOSTROPHE:
+   * Regex that allows only:
+   * - a-z or A-Z characters.
+   * - an empty space.
+   * - a hyphen
+   * - an apostrophe
+   */
+  ALPHA_CHARACTERS_AND_SPACE_HYPHEN_APOSTROPHE: /^[A-Z \-']*$/i,
```
Comment (ttbarnes) [inline src/ui/server/constants/regex.ts:18]: This is exactly the same as `ALPHA_CHARACTERS_AND_SPACE`?

File: src/ui/server/controllers/insurance/account/create/your-details/validation/rules/last-name.ts [hunk 7]
```diff
@@ -21,6 +21,6 @@ const {
  * @param {Object} Errors object from previous validation errors
  * @returns {ValidationErrors}
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/account/create/your-details/validation/rules/last-name.ts:22]: `* @returns {ValidationErrors} nameValidation`

File: src/ui/server/controllers/insurance/account/create/your-details/validation/rules/first-name.ts [hunk 8]
```diff
@@ -21,6 +21,6 @@ const {
  * @param {Object} Errors object from previous validation errors
  * @returns {ValidationErrors}
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/account/create/your-details/validation/rules/first-name.ts:22]: `* @returns {ValidationErrors} nameValidation`

File: src/ui/server/controllers/insurance/policy/name-on-policy/validation/rules/position.ts [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import INSURANCE_FIELD_IDS from '../../../../../../constants/field-ids/insurance';
 import { ERROR_MESSAGES } from '../../../../../../content-strings';
 import { POLICY_FIELDS } from '../../../../../../content-strings/fields/insurance/policy';
-import alphaCharactersAndMaxLengthValidation from '../../../../../../shared-validation/alpha-characters-and-max-length';
+import nameValidation from '../../../../../../shared-validation/name';
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/policy/name-on-policy/validation/rules/position.ts:4]: Is this expected, as IMO i'd say "position" is not a name field?

File: src/ui/server/shared-validation/alpha-characters-hyphen-and-apostrophe-only/index.test.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import alphaCharactersHyphenAndApostropheOnlyValidation from '.';
+import regexValidation from '../regex-validation';
+import { mockErrors, mockErrorMessagesObject } from '../../test-mocks';
+import { REGEX } from '../../constants';
+
+describe('shared-validation/alpha-characters-hyphen-and-apostrophe-only', () => {
+  const mockFieldId = 'Mock field';
+
+  describe('when the value has special characters', () => {
```
Comment (ttbarnes) [inline src/ui/server/shared-validation/alpha-characters-hyphen-and-apostrophe-only/index.test.ts:9]: I don't think we need the `describe` here as the function just calls `regexValidation` regardless?

File: src/ui/server/shared-validation/alpha-characters-only/index.test.ts [hunk 11]
```diff
@@ -1,55 +1,20 @@
...
 const mockFieldId = 'mockId';
-const mockErrorMessage = 'Mock error';
-const mockErrors = {};
-const mockSpecialCharacters = '!@£$%^&*()?';
 
 describe('shared-validation/alpha-characters-only', () => {
   describe('when a string contains numbers', () => {
-    it('should return generateValidationErrors', () => {
-      const mockFieldValue = 'mock 123';
+    it('should return the result of regexValidation', () => {
```
Comment (ttbarnes) [inline src/ui/server/shared-validation/alpha-characters-only/index.test.ts:10]: Same as above, can remove the `describe`

File: src/ui/server/shared-validation/alpha-characters-only/index.ts [hunk 12]
```diff
@@ -11,20 +10,7 @@ import generateValidationErrors from '../../helpers/validation';
...
-  const validation = schema().validate(fieldValue);
-
-  if (validation.error) {
-    return generateValidationErrors(fieldId, errorMessage, errors);
-  }
-
-  return false;
-};
+const alphaCharactersOnlyValidation = (fieldValue: string, fieldId: string, errorMessage: string, errors: object) =>
+  regexValidation(fieldValue, fieldId, REGEX.ALPHA_CHARACTERS_AND_SPACE, errorMessage, errors);
```
Comment (ttbarnes) [inline src/ui/server/shared-validation/alpha-characters-only/index.ts:14]: Could we update `numberHyphenSpacesOnlyValidation` to use `regexValidation? Then all instances of `joiString.regex(...` are aligned

File: src/ui/server/controllers/insurance/policy/name-on-policy/validation/rules/position.ts (hunk 9 above)
Comment (Zainzzkk) [reply src/ui/server/controllers/insurance/policy/name-on-policy/validation/rules/position.ts:4]: it uses the same validation.  Would be duplication if I made one for position

File: src/ui/server/shared-validation/regex-validation/index.ts [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+import generateValidationErrors from '../../helpers/validation';
+
+/**
+ * regexValidation
+ * Check that a string matches a regex pattern
+ * @param {String} fieldValue: Field value to assert
+ * @param {String} fieldId: Field ID
+ * @param {String} errorMessage: Error message to use if invalid
+ * @param {Object} Validation errors
+ * @returns {Object | Boolean} Validation errors
```
Comment (ttbarnes) [inline src/ui/server/shared-validation/regex-validation/index.ts:11]: Missing `regex` documentation

Can we add `errorMessage` to line 10 as well?

**Review by ttbarnes — APPROVED** [review]

**Pull request merged**